geomstats/geometry/poincare_ball.py

```diff
@@ -305,3 +437,252 @@ def inner_product_matrix(self, base_point=None):
+        val = gs.expand_dims(variance, 1)
+        val = gs.repeat(val, precision, axis=1)
+
+        difference = gs.abs(ref - val)
+
+        index = gs.argmin(difference, axis=-1)
+        result = self.normalization_factor_var[index]
+
+        return result
+
+    def find_variance_from_index(self, weighted_distances):
+        """Return the variance given weighted distances.
+
+        Parameters
+        ----------
+        weighted_distances : array-like, shape=[n_gaussian,]
+            Mean of the weighted distances between training data
+            and current barycentres. The weights of each data sample
+            corresponds to the probability of belonging to a component
+            of the Gaussian mixture model.
+
+        Returns
+        -------
+        result : array-like, shape=[n_gaussian]
+            Estimated variances for each component of the GMM.
+        """
+        n_gaussian, precision = \
+            weighted_distances.shape[0], self.variances.shape[0]
+
+        ref = gs.expand_dims(self.phi_inv_var, 0)
+        ref = gs.repeat(ref, n_gaussian, axis=0)
+
+        val = gs.expand_dims(weighted_distances, 1)
+        val = gs.repeat(val, precision, axis=1)
+
+        abs_difference = gs.abs(ref - val)
+
+        index = gs.argmin(abs_difference, -1)
+
+        result = self.variances[index]
+
+        return result
+
+    def normalization_factor(self, variances):
+        """Return normalization factor.
+
+        Parameters
+        ----------
+        variances : array-like, shape=[n,]
+            An array of equally distant values of the
+            variance precision time.
+
+        Returns
+        -------
+        norm_func : array-like, shape=[n,]
+            Normalisation factor for all given variances.
+        """
+        binomial_coefficient = None
+        n_samples = variances.shape[0]
+
+        expand_variances = gs.expand_dims(variances, axis=0)
+        expand_variances = gs.repeat(expand_variances, self.dim, axis=0)
+
+        if binomial_coefficient is None:
+
+            dim_range = gs.arange(self.dim)
+            dim_range[0] = 1
+            n_fact = dim_range.prod()
+
+            k_fact = gs.concatenate(
+                [gs.expand_dims(dim_range[:i].prod(), 0)
+                 for i in range(1, dim_range.shape[0] + 1)], 0)
+
+            nmk_fact = gs.flip(k_fact, 0)
+
+            binomial_coefficient = n_fact / (k_fact * nmk_fact)
+
+        binomial_coefficient = gs.expand_dims(binomial_coefficient, -1)
+        binomial_coefficient = gs.repeat(binomial_coefficient,
+                                         n_samples, axis=1)
+
+        range_ = gs.expand_dims(gs.arange(self.dim), -1)
+        range_ = gs.repeat(range_, n_samples, axis=1)
+
+        ones_ = gs.expand_dims(gs.ones(self.dim), -1)
+        ones_ = gs.repeat(ones_, n_samples, axis=1)
+
+        alternate_neg = (-ones_) ** (range_)
+
+        ins_gs = (((self.dim - 1) - 2 * range_) *
```

Comment:
what do ins_gs, as_o_gs, bs_o_gs, and r_gs mean? Maybe more explicit variable names could help understand this part?